docs/Execution.md

```diff
@@ -0,0 +1,107 @@
+
+typedef
+_IRQL_requires_max_(PASSIVE_LEVEL)
+_Function_class_(QUIC_LISTENER_CALLBACK)
+QUIC_STATUS
+(QUIC_API QUIC_LISTENER_CALLBACK)(
+    _In_ HQUIC Listener,
+    _In_opt_ void* Context,
+    _Inout_ QUIC_LISTENER_EVENT* Event
+    );
+```
+
+Above is an example of a callback delivered to the listener interface.
+The application must register a callback handler to manage all events MsQuic may indicate, returning a status to show if it was successfully handled or not.
+
+This approach differs significantly from sockets and most networking libraries, where the application must make a call (e.g., `send` or `recv`) to determine if something happened.
+This design choice was made for several reasons:
+
+- The MsQuic API **runs in-process**, eliminating the need for a kernel to user mode boundary switch to notify the application layer. This makes the callback-based design more practical compared to sockets.
+
+- MsQuic, due to the QUIC protocol, has numerous event types. Applications may have hundreds of objects with potential state changes. The callback model allows the application to avoid managing pending calls on each object.
+
+- Writing correct, scalable  code on top of the socket interfaces has proven challenging. Callbacks, executed at the correct time and on the correct thread/processor, enable MsQuic to abstract much complexity from applications, making things "just work" out of the box.
+
+- It simplifies MsQuic's logic by eliminating the need for a queue or cached state to indicate to the application. In the socket model, the networking stack must wait for a top-down call from the application before indicating completion, increasing code size, complexity, and memory usage.
+
+### Writing Event Handlers
+
+Event handlers are **essential** for all objects that support them, as much of the MsQuic API operates through these callbacks.
+Critical events, such as "shutdown complete" notifications, provide vital information necessary for the application to function correctly.
+Without these events, the application cannot determine when it is safe to clean up objects.
+
+Applications should keep the execution time within callbacks **to a minimum**.
+MsQuic does not use separate threads for protocol execution and upcalls to the application.
+Therefore, any significant delays in the callback **will delay the protocol**.
+Any substantial work required by the application should be performed on its own thread.
+
+This does not mean the application cannot perform any work in the callback handler.
+In fact, many operations are designed to be most efficient when executed within the callback.
+For example, closing a handle to a connection or stream is ideally done during the "shutdown complete" indication.
+
+A crucial aspect of this design is that all blocking API (down) calls invoked within a callback always occur inline (to prevent deadlocks) and will take precedence over any calls in progress or queued from a separate thread.
```

Comment:
does QUIC make any promises about recursive callbacks? e.g. if an app calls QuicDoFoo() from within the first MsQuic callback, could MsQuic invoke a FooComplete() callback? if not, does this imply MsQuic in fact builds a queue of notifications, contrary to the point above?

Reply:
MsQuic can recursively invoke callbacks in some limited cases, but only if the app opts in. But most of the time, there is no need.

Reply:
should that option (and default behavior) be described here? this is a common question for any code interacting with callbacks, because it can significantly affect implementations, e.g., can i acquire a spinlock within a callback?

Reply:
Sure. I will add a line about recursive callback behavior.